File: libs/schemas/experiment.py
"""Phase 3 Pydantic schemas for hypotheses, protocols, execution, and verification."""

from __future__ import annotations

from datetime import datetime
from typing import Any
from uuid import UUID

from pydantic import BaseModel, Field

# ---------------------------------------------------------------------------
# Hypothesis session
# ---------------------------------------------------------------------------


class HypothesisBudget(BaseModel):
    """Budget constraints for hypothesis generation."""

    max_hypotheses: int = Field(default=10, ge=1, le=50)
    max_critique_rounds: int = Field(default=2, ge=1, le=5)


class HypothesisSessionStartRequest(BaseModel):
    """Request to start a hypothesis generation session."""

    cycle_id: UUID
    charter_id: UUID
    budget: HypothesisBudget = Field(default_factory=HypothesisBudget)


class HypothesisSessionRead(BaseModel):
    """Read model for a hypothesis session."""

    model_config = {"from_attributes": True}

    id: UUID
    cycle_id: UUID
    charter_id: UUID
    status: str
    budget: dict[str, Any] | None = None
    stats: dict[str, Any] | None = None
    step_log: list[Any] | None = None
    error: str | None = None
    created_at: datetime
    updated_at: datetime
    started_at: datetime | None = None
    completed_at: datetime | None = None


class HypothesisSessionStartResponse(BaseModel):
    """Response after starting a hypothesis session."""

    session: HypothesisSessionRead
    job_id: UUID


# ---------------------------------------------------------------------------
# Hypothesis card
# ---------------------------------------------------------------------------


class HypothesisCardRead(BaseModel):
    """Read model for a hypothesis card."""

    model_config = {"from_attributes": True}

    id: UUID
    hypothesis_session_id: UUID
    charter_id: UUID
    cycle_id: UUID
    title: str
    statement: str
    rationale: str
    mechanism: str | None = None
    supporting_evidence_ids: list[str]
    critique: dict[str, Any] | None = None
    novelty_score: float | None = None
    feasibility_score: float | None = None
    impact_score: float | None = None
    rank: int | None = None
    status: str
    rejection_reason: str | None = None
    created_at: datetime
    updated_at: datetime


class HypothesisCardUpdate(BaseModel):
    """Update model for a hypothesis card (select/reject/defer)."""

    status: str | None = Field(
        default=None,
        pattern="^(candidate|selected|rejected|deferred)$",
    )
    rejection_reason: str | None = None


# ---------------------------------------------------------------------------
# Experiment spec
# ---------------------------------------------------------------------------


class ExperimentSpecCompileRequest(BaseModel):
    """Request to compile protocols for selected hypotheses."""

    cycle_id: UUID
    charter_id: UUID
    hypothesis_card_ids: list[UUID] | None = Field(
        default=None,
        description="Specific cards to compile. If None, compiles top-ranked.",
    )
    hardware_profile: dict[str, Any] | None = None
    base_image: str | None = None


class ExperimentSpecRead(BaseModel):
    """Read model for an experiment spec."""

    model_config = {"from_attributes": True}

    id: UUID
    hypothesis_card_id: UUID
    charter_id: UUID
    cycle_id: UUID
    title: str
    description: str
    baseline: dict[str, Any]
    controls: list[dict[str, Any]]
    metrics: list[dict[str, Any]]
    expected_artifacts: list[dict[str, Any]]
    stop_conditions: list[dict[str, Any]]
    code_plan: dict[str, Any] | None = None
    hardware_profile: dict[str, Any] | None = None
    base_image: str | None = None
    build_recipe: dict[str, Any] | None = None
    status: str
    rejection_reason: str | None = None
    created_at: datetime
    updated_at: datetime


class ExperimentSpecCompileResponse(BaseModel):
    """Response after compiling protocols."""

    specs: list[ExperimentSpecRead]
    job_id: UUID


# ---------------------------------------------------------------------------
# Run record
# ---------------------------------------------------------------------------


class RunStartRequest(BaseModel):
    """Request to start an experiment run."""

    experiment_spec_id: UUID
    gpu_enabled: bool = False


class RunControlRequest(BaseModel):
    """Request to control a running experiment."""

    action: str = Field(pattern="^(pause|resume|cancel|retry)$")


class RunRecordRead(BaseModel):
    """Read model for a run record."""

    model_config = {"from_attributes": True}

    id: UUID
    experiment_spec_id: UUID
    charter_id: UUID
    cycle_id: UUID
    run_number: int
    status: str
    workspace_path: str | None = None
    container_id: str | None = None
    image_ref: str | None = None
    command: str | None = None
    env_vars: dict[str, Any] | None = None
    resource_limits: dict[str, Any] | None = None
    exit_code: int | None = None
    stdout_path: str | None = None
    stderr_path: str | None = None
    metrics_output: dict[str, Any] | None = None
    artifact_manifest: list[dict[str, Any]] | None = None
    resource_usage: dict[str, Any] | None = None
    failure_class: str | None = None
    error: str | None = None
    started_at: datetime | None = None
    completed_at: datetime | None = None
    created_at: datetime
    updated_at: datetime


class RunStartResponse(BaseModel):
    """Response after starting a run."""

    run: RunRecordRead
    job_id: UUID


# ---------------------------------------------------------------------------
# Run telemetry
# ---------------------------------------------------------------------------


class RunTelemetryRead(BaseModel):
    """Read model for a telemetry row."""

    model_config = {"from_attributes": True}

    id: UUID
    run_record_id: UUID
    timestamp: datetime
    event_type: str
    payload: dict[str, Any]


# ---------------------------------------------------------------------------
# Verification report
# ---------------------------------------------------------------------------


class VerificationReportRead(BaseModel):
    """Read model for a verification report."""

    model_config = {"from_attributes": True}

    id: UUID
    run_record_id: UUID
    charter_id: UUID
    cycle_id: UUID
    verdict: str
    baseline_comparison: dict[str, Any] | None = None
    artifact_checks: list[dict[str, Any]] | None = None
    output_contract: dict[str, Any] | None = None
    metric_sanity: dict[str, Any] | None = None
    warnings: list[str] | None = None
    summary: str
    created_at: datetime


# ---------------------------------------------------------------------------
# Failure postmortem
# ---------------------------------------------------------------------------


class FailurePostmortemRead(BaseModel):
    """Read model for a failure postmortem."""

    model_config = {"from_attributes": True}

    id: UUID
    run_record_id: UUID
    verification_report_id: UUID | None = None
    charter_id: UUID
    cycle_id: UUID
    failure_class: str
    root_cause: str
    contributing_factors: list[dict[str, Any]] | None = None
    error_trace: str | None = None
    next_step_recommendation: str | None = None
    lessons: list[dict[str, Any]] | None = None
    created_at: datetime

File: apps/web/src/api/experiment.ts
// API client for Phase 3 experiment endpoints

import type { PaginatedResponse } from "./client";

const BASE_URL = import.meta.env.VITE_API_BASE_URL ?? "/api/v1";

async function apiFetch<T>(path: string, init?: RequestInit): Promise<T> {
  const url = `${BASE_URL}${path}`;
  const res = await fetch(url, {
    ...init,
    headers: { "Content-Type": "application/json", ...init?.headers },
  });
  if (!res.ok) {
    const body = await res.text().catch(() => "");
    throw new Error(`API ${res.status}: ${body || res.statusText}`);
  }
  return res.json() as Promise<T>;
}

// ---- Types ----

export interface HypothesisSession {
  id: string;
  cycle_id: string;
  charter_id: string;
  status: string;
  budget: Record<string, unknown> | null;
  stats: Record<string, unknown> | null;
  step_log: Array<Record<string, unknown>> | null;
  error: string | null;
  created_at: string;
  updated_at: string;
  started_at: string | null;
  completed_at: string | null;
}

export interface HypothesisCard {
  id: string;
  hypothesis_session_id: string;
  charter_id: string;
  cycle_id: string;
  title: string;
  statement: string;
  rationale: string;
  mechanism: string | null;
  supporting_evidence_ids: string[];
  critique: Record<string, unknown> | null;
  novelty_score: number | null;
  feasibility_score: number | null;
  impact_score: number | null;
  rank: number | null;
  status: string;
  rejection_reason: string | null;
  created_at: string;
  updated_at: string;
}

export interface ExperimentSpec {
  id: string;
  hypothesis_card_id: string;
  charter_id: string;
  cycle_id: string;
  title: string;
  description: string;
  baseline: Record<string, unknown>;
  controls: Array<Record<string, unknown>>;
  metrics: Array<Record<string, unknown>>;
  expected_artifacts: Array<Record<string, unknown>>;
  stop_conditions: Array<Record<string, unknown>>;
  code_plan: Record<string, unknown> | null;
  hardware_profile: Record<string, unknown> | null;
  base_image: string | null;
  status: string;
  rejection_reason: string | null;
  created_at: string;
  updated_at: string;
}

export interface RunRecord {
  id: string;
  experiment_spec_id: string;
  parent_run_id: string | null;
  charter_id: string;
  cycle_id: string;
  run_number: number;
  status: string;
  workspace_path: string | null;
  container_id: string | null;
  image_ref: string | null;
  command: string | null;
  exit_code: number | null;
  metrics_output: Record<string, number> | null;
  artifact_manifest: Array<Record<string, unknown>> | null;
  resource_usage: Record<string, unknown> | null;
  failure_class: string | null;
  error: string | null;
  started_at: string | null;
  completed_at: string | null;
  created_at: string;
  updated_at: string;
}

export interface RunTelemetry {
  id: string;
  run_record_id: string;
  timestamp: string;
  event_type: string;
  payload: Record<string, unknown>;
}

export interface VerificationReport {
  id: string;
  run_record_id: string;
  charter_id: string;
  cycle_id: string;
  verdict: string;
  baseline_comparison: Record<string, unknown> | null;
  artifact_checks: Array<Record<string, unknown>> | null;
  warnings: string[] | null;
  summary: string;
  created_at: string;
}

export interface FailurePostmortem {
  id: string;
  run_record_id: string;
  charter_id: string;
  cycle_id: string;
  failure_class: string;
  root_cause: string;
  contributing_factors: Array<Record<string, unknown>> | null;
  error_trace: string | null;
  next_step_recommendation: string | null;
  lessons: Array<Record<string, unknown>> | null;
  created_at: string;
}

export interface HypothesisSessionStartResponse {
  session: HypothesisSession;
  job_id: string;
}

export interface ExperimentSpecCompileResponse {
  specs: ExperimentSpec[];
  job_id: string;
}

export interface RunStartResponse {
  run: RunRecord;
  job_id: string;
}

// ---- Hypothesis API ----

export function startHypothesisSession(body: {
  cycle_id: string;
  charter_id: string;
  budget?: { max_hypotheses?: number; max_critique_rounds?: number };
}) {
  return apiFetch<HypothesisSessionStartResponse>(
    "/hypotheses/sessions",
    { method: "POST", body: JSON.stringify(body) },
  );
}

export function fetchHypothesisSessions(opts: {
  charter_id?: string;
  cycle_id?: string;
  offset?: number;
  limit?: number;
} = {}) {
  const params = new URLSearchParams();
  if (opts.charter_id) params.set("charter_id", opts.charter_id);
  if (opts.cycle_id) params.set("cycle_id", opts.cycle_id);
  if (opts.offset !== undefined) params.set("offset", String(opts.offset));
  if (opts.limit !== undefined) params.set("limit", String(opts.limit));
  const qs = params.toString();
  return apiFetch<PaginatedResponse<HypothesisSession>>(
    `/hypotheses/sessions${qs ? `?${qs}` : ""}`,
  );
}

export function fetchHypothesisSession(sessionId: string) {
  return apiFetch<HypothesisSession>(`/hypotheses/sessions/${sessionId}`);
}

export function fetchHypothesisCards(opts: {
  cycle_id?: string;
  status?: string;
  offset?: number;
  limit?: number;
} = {}) {
  const params = new URLSearchParams();
  if (opts.cycle_id) params.set("cycle_id", opts.cycle_id);
  if (opts.status) params.set("status", opts.status);
  if (opts.offset !== undefined) params.set("offset", String(opts.offset));
  if (opts.limit !== undefined) params.set("limit", String(opts.limit));
  const qs = params.toString();
  return apiFetch<PaginatedResponse<HypothesisCard>>(
    `/hypotheses/cards${qs ? `?${qs}` : ""}`,
  );
}

export function fetchHypothesisCard(cardId: string) {
  return apiFetch<HypothesisCard>(`/hypotheses/cards/${cardId}`);
}

export function updateHypothesisCard(
  cardId: string,
  body: { status?: string; rejection_reason?: string },
) {
  return apiFetch<HypothesisCard>(`/hypotheses/cards/${cardId}`, {
    method: "PATCH",
    body: JSON.stringify(body),
  });
}

// ---- Protocol API ----

export function compileProtocols(body: {
  cycle_id: string;
  charter_id: string;
  hypothesis_card_ids?: string[];
  hardware_profile?: Record<string, unknown>;
  base_image?: string;
}) {
  return apiFetch<ExperimentSpecCompileResponse>(
    "/protocols/compile",
    { method: "POST", body: JSON.stringify(body) },
  );
}

export function fetchExperimentSpecs(opts: {
  cycle_id?: string;
  status?: string;
  offset?: number;
  limit?: number;
} = {}) {
  const params = new URLSearchParams();
  if (opts.cycle_id) params.set("cycle_id", opts.cycle_id);
  if (opts.status) params.set("status", opts.status);
  if (opts.offset !== undefined) params.set("offset", String(opts.offset));
  if (opts.limit !== undefined) params.set("limit", String(opts.limit));
  const qs = params.toString();
  return apiFetch<PaginatedResponse<ExperimentSpec>>(
    `/protocols/specs${qs ? `?${qs}` : ""}`,
  );
}

export function fetchExperimentSpec(specId: string) {
  return apiFetch<ExperimentSpec>(`/protocols/specs/${specId}`);
}

// ---- Run API ----

export function startRun(body: {
  experiment_spec_id: string;
  gpu_enabled?: boolean;
}) {
  return apiFetch<RunStartResponse>(
    "/runs",
    { method: "POST", body: JSON.stringify(body) },
  );
}

export function fetchRunRecords(opts: {
  cycle_id?: string;
  spec_id?: string;
  status?: string;
  offset?: number;
  limit?: number;
} = {}) {
  const params = new URLSearchParams();
  if (opts.cycle_id) params.set("cycle_id", opts.cycle_id);
  if (opts.spec_id) params.set("spec_id", opts.spec_id);
  if (opts.status) params.set("status", opts.status);
  if (opts.offset !== undefined) params.set("offset", String(opts.offset));
  if (opts.limit !== undefined) params.set("limit", String(opts.limit));
  const qs = params.toString();
  return apiFetch<PaginatedResponse<RunRecord>>(
    `/runs${qs ? `?${qs}` : ""}`,
  );
}

export function fetchRunRecord(runId: string) {
  return apiFetch<RunRecord>(`/runs/${runId}`);
}

export function controlRun(runId: string, action: string) {
  return apiFetch<RunRecord>(`/runs/${runId}/control`, {
    method: "POST",
    body: JSON.stringify({ action }),
  });
}

export function fetchRunTelemetry(runId: string, opts: {
  since?: string;
  limit?: number;
} = {}) {
  const params = new URLSearchParams();
  if (opts.since) params.set("since", opts.since);
  if (opts.limit !== undefined) params.set("limit", String(opts.limit));
  const qs = params.toString();
  return apiFetch<RunTelemetry[]>(
    `/runs/${runId}/telemetry${qs ? `?${qs}` : ""}`,
  );
}

// ---- Verification API ----

export function fetchVerificationReport(runId: string) {
  return apiFetch<VerificationReport>(`/runs/${runId}/verification`);
}

export function fetchFailurePostmortem(runId: string) {
  return apiFetch<FailurePostmortem>(`/runs/${runId}/postmortem`);
}

export function fetchVerificationReports(opts: {
  cycle_id?: string;
  verdict?: string;
  offset?: number;
  limit?: number;
} = {}) {
  const params = new URLSearchParams();
  if (opts.cycle_id) params.set("cycle_id", opts.cycle_id);
  if (opts.verdict) params.set("verdict", opts.verdict);
  if (opts.offset !== undefined) params.set("offset", String(opts.offset));
  if (opts.limit !== undefined) params.set("limit", String(opts.limit));
  const qs = params.toString();
  return apiFetch<PaginatedResponse<VerificationReport>>(
    `/verifications${qs ? `?${qs}` : ""}`,
  );
}
